space_packet_parser/packets.py

```diff
@@ -29,13 +30,67 @@ class SequenceFlags(IntEnum):
 
 
 class RawPacketData(bytes):
-    """A class to represent raw packet data as bytes but whose length is represented by bit length.
+    """A class containing the raw binary packet data.
 
     This class is a subclass of bytes and is used to represent the raw packet data
-    in a more readable way. It is used to store the raw packet data in the Packet
+    in a more readable way. It is used to store the binary data in the Packet
     class and used to keep track of the current parsing position (accessible through the `pos` attribute).
+    """
+    pos = 0  # Current position in bits
+
+    def read_as_bytes(self, nbits: int) -> bytes:
+        """Read a number of bits from the packet data as bytes. Reads minimum number of complete bytes required to
+        capture `nbits`. Moves `pos` cursor `nbits` forward, even if `nbits` is not an integer number of bytes.
+
+        Parameters
+        ----------
+        nbits : int
+            Number of bits to read
+
+        Returns
+        -------
+        : bytes
+            Raw bytes from the packet data
+        """
+        if self.pos + nbits > len(self) * 8:
+            raise ValueError("Tried to read beyond the end of the packet data. "
+                             f"Tried to read {nbits} bits from position {self.pos} "
+                             f"in a packet of length {len(self) * 8} bits.")
+        if self.pos % 8 == 0 and nbits % 8 == 0:
+            # If the read is byte-aligned, we can just return the bytes directly
+            data = self[self.pos//8:self.pos//8 + (nbits+7) // 8]
+            self.pos += nbits
+            return data
+        # We are non-byte aligned, so we need to extract the bits and convert to bytes
+        bytes_as_int = _extract_bits(self, self.pos, nbits)
+        self.pos += nbits
+        return int.to_bytes(bytes_as_int, (nbits + 7) // 8, "big")
+
+    def read_as_int(self, nbits: int) -> int:
+        """Read a number of bits from the packet data as an integer.
+
+        Parameters
+        ----------
+        nbits : int
+            Number of bits to read
 
-    There are also a few convenience methods to extract the CCSDS header fields from the packet data.
+        Returns
+        -------
+        : int
+            Integer representation of the bits read from the packet
+        """
+        if self.pos + nbits > len(self) * 8:
+            raise ValueError("Tried to read beyond the end of the packet data. "
+                             f"Tried to read {nbits} bits from position {self.pos} "
+                             f"in a packet of length {len(self) * 8} bits.")
+        int_data = _extract_bits(self, self.pos, nbits)
+        self.pos += nbits
+        return int_data
+
+class CCSDSPacketBytes(bytes):
+    """Binary representation of a CCSDS packet.
+
+    Methods to extract the header fields are added to the raw bytes object.
     """
     HEADER_LENGTH_BYTES = 6
     pos = 0  # in bits
```

Comment:
Would it be appropriate to add bookkeeping to the `CCSDSPacketBytes` class to keep track of the necessary information for concatenating and parsing continued packets? The signature of the class initializer wouldn't need to change but we could add instance attributes that mark it as a continued packet, how many packets are continued, and how big the secondary header is. The idea would be to self contain all the necessary information needed to correctly parse it later.

Reply:
I'm not sure I follow what you mean by correctly parse it later? The XTCE definition should not contain any information about the inbetween headers, so those bytes are meant to be dropped in the CCSDS spec.

If a user wants that information then they should call this function without concatenating the data together I think. I don't see it being super useful personally to bookkeep that data because there is a way to get it by going back to the source.